File: silentpilot/mcp_server/src/tools/browser.ts
/**
 * Playwright-backed browser tools for the MCP server.
 *
 * 8 tools intentionally kept small for model reliability:
 * - browser_goto: navigate to URL
 * - browser_click: click an element by selector
 * - browser_type: type text into an element
 * - browser_press: press a keyboard key
 * - browser_scroll: scroll the page
 * - browser_screenshot: capture viewport as base64 PNG
 * - browser_extract_text: get visible text content
 * - browser_wait: wait for a duration
 */

import { chromium, Browser, Page } from "playwright";
import { PageCursor } from "./cursor.js";

let browser: Browser | null = null;
let page: Page | null = null;
let cursor: PageCursor | null = null;

const VIEWPORT = { width: 1280, height: 800 };

export async function ensureBrowser(): Promise<Page> {
  if (!page || !browser?.isConnected()) {
    if (browser) {
      try { await browser.close(); } catch {}
    }

    const headless = process.env.BROWSER_HEADLESS !== "false";

    browser = await chromium.launch({
      headless,
      args: [
        "--disable-extensions",
        "--disable-file-system",
        "--no-sandbox",
      ],
    });

    const context = await browser.newContext({
      viewport: VIEWPORT,
    });
    page = await context.newPage();
    cursor = new PageCursor(page);
    await cursor.attach();
  } else {
    // Ensure cursor is still in DOM (handles navigations)
    if (cursor) await cursor.ensureAlive();
  }
  return page;
}

export async function closeBrowser(): Promise<void> {
  if (browser) {
    await browser.close();
    browser = null;
    page = null;
  }
}

// --- Tool Implementations ---

export async function browserGoto(url: string): Promise<string> {
  const p = await ensureBrowser();
  if (cursor) await cursor.hide();
  await p.goto(url, { waitUntil: "domcontentloaded", timeout: 30000 });
  if (cursor) {
    await cursor.attach();
    await cursor.show();
  }
  return `Navigated to ${p.url()}`;
}

export async function browserClick(selector: string): Promise<string> {
  const p = await ensureBrowser();
  try {
    const loc = p.locator(selector);
    const box = await loc.boundingBox();
    if (box && cursor) {
      await cursor.moveTo(box.x + box.width / 2, box.y + box.height / 2);
      await cursor.clickEffect();
    }
    await loc.click({ timeout: 5000 });
    return `Clicked: ${selector}`;
  } catch (e) {
    return `Failed to click '${selector}': ${e}`;
  }
}

export async function browserType(
  selector: string,
  text: string
): Promise<string> {
  const p = await ensureBrowser();
  try {
    await p.fill(selector, text, { timeout: 5000 });
    return `Typed into ${selector}: "${text}"`;
  } catch (e) {
    return `Failed to type into '${selector}': ${e}`;
  }
}

export async function browserPress(key: string): Promise<string> {
  const p = await ensureBrowser();
  await p.keyboard.press(key);
  return `Pressed key: ${key}`;
}

export async function browserScroll(
  direction: "up" | "down",
  amount: number = 300
): Promise<string> {
  const p = await ensureBrowser();
  if (cursor) {
    await cursor.moveTo(VIEWPORT.width / 2, VIEWPORT.height / 2);
  }
  const delta = direction === "down" ? amount : -amount;
  await p.mouse.wheel(0, delta);
  return `Scrolled ${direction} by ${amount}px`;
}

export async function browserScreenshot(): Promise<string> {
  const p = await ensureBrowser();
  const buffer = await p.screenshot({ type: "png" });
  return buffer.toString("base64");
}

export async function browserExtractText(): Promise<string> {
  const p = await ensureBrowser();
  const text = await p.evaluate(() => {
    return document.body?.innerText?.substring(0, 5000) || "";
  });
  return text;
}

export async function browserWait(ms: number): Promise<string> {
  const p = await ensureBrowser();
  await p.waitForTimeout(Math.min(ms, 10000)); // cap at 10s
  return `Waited ${ms}ms`;
}

File: actions/browser.py
"""Pyppeteer browser controller for taking screenshots and executing actions."""

import asyncio
from pyppeteer import launch

from actions.config import BROWSER_HEADLESS, VIEWPORT_WIDTH, VIEWPORT_HEIGHT, CURSOR_ENABLED
from actions.cursor import PageCursor


class BrowserController:
    """Controls a Chromium browser via Pyppeteer."""

    def __init__(self, on_cursor_move=None):
        self._browser = None
        self._page = None
        self._on_cursor_move = on_cursor_move
        self._cursor: PageCursor | None = None

    async def launch(self):
        """Launch Chromium browser."""
        self._browser = await launch(
            headless=BROWSER_HEADLESS,
            args=[
                f"--window-size={VIEWPORT_WIDTH},{VIEWPORT_HEIGHT}",
                "--no-sandbox",
            ],
        )
        self._page = (await self._browser.pages())[0]
        # Clear fixed viewport so page follows window size
        await self._page._client.send("Emulation.clearDeviceMetricsOverride")

        if CURSOR_ENABLED:
            self._cursor = PageCursor(on_position_change=self._on_cursor_move)
            await self._cursor.attach(self._page)

    async def get_viewport_size(self) -> tuple[int, int]:
        """Get the current actual viewport size from the browser."""
        size = await self._page.evaluate(
            "() => ({w: window.innerWidth, h: window.innerHeight})"
        )
        return int(size["w"]), int(size["h"])

    async def ensure_cursor(self):
        """Re-inject cursor if it was removed by dynamic page updates."""
        if self._cursor:
            await self._cursor.ensure_alive()

    async def screenshot(self) -> bytes:
        """Take a screenshot at current window size."""
        return await self._page.screenshot({"type": "png"})

    async def get_interactive_elements(self) -> list[dict]:
        """Extract all visible interactive elements with bounding boxes.

        Covers: links, buttons, inputs, textareas, selects, role-based elements,
        elements with click handlers, elements with cursor:pointer, and
        any visible text element that looks clickable.
        """
        elements = await self._page.evaluate("""() => {
            const results = [];
            const seen = new Set();
            let id = 0;
            const vw = window.innerWidth;
            const vh = window.innerHeight;

            function getText(el) {
                // Try multiple sources for element text
                return (
                    el.getAttribute('aria-label') ||
                    el.innerText ||
                    el.value ||
                    el.title ||
                    el.placeholder ||
                    el.alt ||
                    el.getAttribute('data-tooltip') ||
                    ''
                ).trim().substring(0, 80);
            }

            function isVisible(el) {
                const style = window.getComputedStyle(el);
                if (style.display === 'none' || style.visibility === 'hidden' || style.opacity === '0') return false;
                const rect = el.getBoundingClientRect();
                if (rect.width < 3 || rect.height < 3) return false;
                if (rect.bottom < 0 || rect.top > vh || rect.right < 0 || rect.left > vw) return false;
                return true;
            }

            function addElement(el) {
                if (id >= 60) return;
                if (!isVisible(el)) return;
                const rect = el.getBoundingClientRect();
                const cx = Math.round(rect.left + rect.width / 2);
                const cy = Math.round(rect.top + rect.height / 2);

                // Deduplicate by center position (within 5px)
                const key = `${Math.round(cx/5)*5},${Math.round(cy/5)*5}`;
                if (seen.has(key)) return;

                const text = getText(el);
                // Allow inputs/textareas even without text
                const isInput = el.tagName === 'INPUT' || el.tagName === 'TEXTAREA' || el.tagName === 'SELECT';
                if (!text && !isInput) return;

                seen.add(key);
                results.push({
                    id: id++,
                    tag: el.tagName.toLowerCase(),
                    text: text || `(${el.type || el.tagName.toLowerCase()} field)`,
                    href: el.href || el.closest('a')?.href || '',
                    type: el.type || '',
                    cx: cx,
                    cy: cy,
                    w: Math.round(rect.width),
                    h: Math.round(rect.height)
                });
            }

            // 1. Standard interactive elements
            const standardSelectors = [
                'a[href]',
                'button',
                'input',
                'textarea',
                'select',
                '[role="button"]',
                '[role="link"]',
                '[role="tab"]',
                '[role="menuitem"]',
                '[role="option"]',
                '[role="checkbox"]',
                '[role="radio"]',
                '[role="switch"]',
                '[role="combobox"]',
                '[role="searchbox"]',
                '[onclick]',
                '[tabindex]:not([tabindex="-1"])',
                'summary',
                'label[for]',
            ].join(', ');
            document.querySelectorAll(standardSelectors).forEach(addElement);

            // 2. Elements with cursor:pointer that weren't caught above
            const allVisible = document.querySelectorAll('div, span, li, img, svg, p, h1, h2, h3, td, th');
            for (const el of allVisible) {
                if (id >= 60) break;
                const style = window.getComputedStyle(el);
                if (style.cursor === 'pointer') {
                    addElement(el);
                }
            }

            return results;
        }""")
        return elements

    async def click(self, selector: str):
        """Click an element by CSS selector."""
        await self._page.click(selector)

    async def click_coords(self, x: int, y: int):
        """Click at specific pixel coordinates."""
        if self._cursor:
            await self._cursor.move_to(x, y)
            await self._cursor.click_effect()
        await self._page.mouse.click(x, y)

    async def scroll(self, direction: str):
        """Scroll the page up or down with smooth animation."""
        if self._cursor:
            w, h = await self.get_viewport_size()
            await self._cursor.move_to(w // 2, h // 2, duration_ms=200)
        delta = -800 if direction == "up" else 800
        await self._page.evaluate(f"window.scrollBy({{top: {delta}, behavior: 'smooth'}})")

    async def goto(self, url: str):
        """Navigate to a URL and wait for page to be interactive."""
        try:
            await self._page.goto(url, {"waitUntil": "networkidle2", "timeout": 12000})
        except Exception:
            pass
        # Extra buffer for late-loading JS content
        await asyncio.sleep(0.5)
        if self._cursor:
            await self._cursor._inject_cursor()
            await self._cursor._center()

    async def type_text(self, selector: str, text: str):
        """Type text into an input element."""
        await self._page.type(selector, text)

    async def page_type(self, text: str):
        """Type text into whatever element is currently focused."""
        await self._page.keyboard.type(text)

    async def press_key(self, key: str):
        """Press a keyboard key (e.g. 'Enter', 'Tab')."""
        await self._page.keyboard.press(key)

    async def go_back(self):
        """Go back in browser history."""
        await self._page.goBack({"waitUntil": "domcontentloaded", "timeout": 8000})

    async def go_forward(self):
        """Go forward in browser history."""
        await self._page.goForward({"waitUntil": "domcontentloaded", "timeout": 8000})

    async def get_url(self) -> str:
        """Get the current page URL."""
        return self._page.url

    async def close(self):
        """Close the browser."""
        if self._browser:
            await self._browser.close()
            self._browser = None
            self._page = None
